mail_tracking_mailgun/models/mail_tracking_email.py

```diff
@@ -0,0 +1,196 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Antonio Espinosa - <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import hashlib
+import hmac
+import threading
+from datetime import datetime
+from openerp import models, api, fields
+
+import logging
+_logger = logging.getLogger(__name__)
+
+
+class MailTrackingEmail(models.Model):
+    _inherit = "mail.tracking.email"
+
+    def _country_search(self, country_code):
+        country = False
+        if country_code:
+            country = self.env['res.country'].search([
+                ('code', '=ilike', country_code),
+            ])
+        if country:
+            return country.id
+        return False
+
+    def _mailgun_mandatory_fields(self):
+        return ('event', 'timestamp', 'token', 'signature',
+                'tracking_email_id', 'odoo_db')
+
+    def _mailgun_event_type_mapping(self):
+        return {
+            # Mailgun event type: tracking event type
+            'delivered': 'delivered',
+            'opened': 'open',
+            'clicked': 'click',
+            'unsubscribed': 'unsub',
+            'complained': 'spam',
+            'bounced': 'hard_bounce',
+            'dropped': 'reject',
+        }
+
+    def _mailgun_supported_event_types(self):
+        return self._mailgun_event_type_mapping().keys()
+
+    def _mailgun_event_type_verify(self, event):
+        event = event or {}
+        mailgun_event_type = event.get('event')
+        if mailgun_event_type not in self._mailgun_supported_event_types():
+            _logger.info("Mailgun: event type '%s' not supported",
+                         mailgun_event_type)
+            return False
+        # OK, event type is valid
+        return True
+
+    def _mailgun_signature(self, api_key, timestamp, token):
+        return hmac.new(
+            key=str(api_key),
+            msg='{}{}'.format(str(timestamp), str(token)),
+            digestmod=hashlib.sha256).hexdigest()
+
+    def _mailgun_signature_verify(self, event):
+        event = event or {}
+        api_key = self.env['ir.config_parameter'].get_param('mailgun.apikey')
+        if not api_key:
+            _logger.info("No Mailgun api key configured. "
+                         "Please add 'mailgun.apikey' to System parameters "
+                         "to enable Mailgun authentication webhoook requests. "
+                         "More info at: "
+                         "https://documentation.mailgun.com/user_manual.html"
+                         "#webhooks")
+        else:
+            timestamp = event.get('timestamp')
+            token = event.get('token')
+            signature = event.get('signature')
+            event_digest = self._mailgun_signature(api_key, timestamp, token)
+            if signature != event_digest:
+                _logger.info("Mailgun: Invalid signature '%s' != '%s'",
```

Comment:
It might be good to bubble this as a `_logger.error`. I feel like this would be indicative of some sort of security issue if this is happening to a production server often, and it would be good to have it stand out a bit more.